[Step5] 5단계 - 자동차 경주(리팩터링)

nooblette opened the pull request:
안녕하세요 현구님! 자동차 경주 마지막 미션이네요 😄 

5단계 미션 PR 드립니다. 
수정 사항은
1. 4단계 코드리뷰 내용 반영
2. 5단계 수정사항 반영
  - 핵심 비즈니스 로직을 갖는 객체(`Car.kt`, `Cars.kt`, `Race.kt`, `RaceService.kt`, `RandomNumbers.kt`)를 domain 패키지로 이동하였습니다. 

프로그래밍 요구 사항 중 테스트 하기 힘든 부분(e.g. Random 값 생성 처리)은 이전 리뷰 내용 중에서 `CarTest.kt`와 `RaceServiceTest.kt`에서 숫자를 파라미터로 넘기면서 처리할 수 있도록 수정하면서 해소가 되었다는 생각이 들었는데, 5단계 요건 중 부족한 부분이 있다면 피드백 부탁드립니다. (이전에 리뷰 주신 내용들이 4단계, 5단계를 빠르게 해결하는데 많은 도움이 됐어요! 👍)

자동차 경주 미션 진행하면서 꼼꼼한 리뷰주셔서 감사드립니다 🙇

nooblette committed:
refactor : 4단계 코드리뷰 내용 적용
- 자동차 이름 경계값 테스트 추가
- 우승자 조회 로직을 Race.kt에서 Cars.kt로 변경

nooblette committed:
refactor : 비즈니스 로직을 갖는 객체를 domain 패키지 내부로 이동

Hyeon9mak commented on src/main/kotlin/racingcar/domain/race/Race.kt:
```diff
@@ -0,0 +1,16 @@
+package racingcar.domain.race
+
+import racingcar.domain.car.Cars
+import racingcar.domain.randomnumber.RandomNumbers
+
+class Race(
+    val cars: Cars,
+) {
+    fun run(randomNumbers: RandomNumbers): Cars {
+        // 자동차 경주
+        return Cars(
+            cars.list.mapIndexed
```
비즈니스적으로 중요한 로직이 담긴 코드들이 모두 domain 패키지로 옮겨지면서,
정책 관련된 동작들을 수정하고 싶다면 다른 패키지는 모두 배제하고
domain 패키지 내부만 확인하면 되겠네요 👍

nooblette replied:
네 맞습니다! 프로그래밍 요구사항에 맞춰 domain 패키지로 분리하면서 훨씬 코드가 유지보수하고 파악하기 편해진 것 같아요!

Hyeon9mak commented on src/main/kotlin/racingcar/domain/service/RaceService.kt:
```diff
@@ -1,8 +1,8 @@
+package racingcar.domain.service
 
-import racingcar.car.Cars
-import racingcar.race.Race
-import racingcar.randomnumber.RandomNumbers
+import racingcar.domain.car.Cars
+import racingcar.domain.race.Race
+import racingcar.domain.randomnumber.RandomNumbers
 
 class RaceService(
     val race: Race,
```
반복문을 돌며 출력을 하기 위해 `RaceService` 에 약간의 로직이 남아있는데요,
혹시 이런건 어떨까요?

1. 레이싱을 반복해서 수행한다. 그리고 그 결과를 따로 모아둔다.
2. `RaceService` 를 통해 View 와 협업하면서 모든 결과를 출력한다.

nooblette replied:
출력을 굳이 매번 수행할 필요는 없었네요 😓 의견 주셔서 감사합니다 👍 

`ResultView`에 리뷰주신 내용과 함께 참고해서 `ResultView`를 domain으로 이동하면서,
`RaceService`가 매 단계별 레이싱 결과를 담고 있다가, View와 바로 협업하여 출력하도록 수정하였습니다.

그리고 이런 리팩토링을 거치면서 `RaceService`의 메서드를 호출할때 출력을 위한 `onResult: (Cars) -> Unit` 함수를 파라미터로 전달하지 않아도 되서 훨씬 간결해진 것 같아요 👍

Hyeon9mak commented on src/main/kotlin/racingcar/view/ResultView.kt:
```diff
@@ -1,6 +1,6 @@
 package racingcar.view
 
-import racingcar.car.Cars
+import racingcar.domain.car.Cars
 
 object ResultView {
     private const val PRINT_EXPRESSION = "-"
```
출력이 domain (비즈니스) 로 포함될 수 있을지 고민해보는 것도 좋겠네요!

nooblette replied:
`RaceService`에 리팩토링을 진행하면서 해당 내용도 반영하여 `ResultView.kt` 파일을 `domain`(비즈니스) 패키지 내부에 포함하도록 수정하였습니다 (제가 이해한 내용이 맞을까요? 😅 )

코드도 훨씬 간결해진 것 같아요 👍 

다만 한가지 고민이 되는게, `ResultView`는 단지 출력을 위한 객체라고 생각했고 전달받은 값을 어떤 형태로, 어느 위치(e.g. 파일, 콘솔, 웹 페이지 등)로 출력할지는 자동차 경주랑 직접적인 상관이 없다고 생각했습니다
결국 `ResultView`는 자동차 경주의 결과를 받고 출력하는 역할만 수행하니 _**domain에 포함될 수 있을까?**_ 라는 고민이 되는데, 관련해서 현구님 의견은 어떠실까요?

Hyeon9mak reviewed (changes requested):
안녕하세요 민혁님! 마지막 단계는 큰 코드 수정이 없었죠? 😉 
이전 단계부터 착실히 리팩터링 해주셨기 때문이라고 생각이 듭니다.

아주 간단한 코멘트만 남겼습니다!
마지막까지 화이팅입니다!

nooblette committed:
refactor : 코드리뷰 내용 적용
1. 레이싱 결과를 매번 출력하지 않고, raceResultByStep에 담아두었다가 한번에 출력
2. ResultView를 domain으로 이동, RaceService가 ResultView에 직접 의존하도록 수정

Hyeon9mak commented on src/main/kotlin/racingcar/domain/result/ResultView.kt:
```diff
@@ -1,6 +1,6 @@
-package racingcar.view
+package racingcar.domain.result
 
-import racingcar.car.Cars
+import racingcar.domain.car.Cars
 
 object ResultView {
     private const val PRINT_EXPRESSION = "-"
```
>  한가지 고민이 되는게, ResultView는 단지 출력을 위한 객체라고 생각했고 전달받은 값을 어떤 형태로, 어느 위치(e.g. 파일, 콘솔, 웹 페이지 등)로 출력할지는 자동차 경주랑 직접적인 상관이 없다고 생각했습니다
결국 ResultView는 자동차 경주의 결과를 받고 출력하는 역할만 수행하니 domain에 포함될 수 있을까? 라는 고민이 되는데, 관련해서 현구님 의견은 어떠실까요?

게임의 결과를 모아둔다는 의미의 `GameResult` 와 같은 객체라면 domain 패키지에 포함될 수 있다고 생각합니다.
게임의 결과는 우리 비즈니스에서 꼭 고려해야할 중요한 사항이고 기능이거든요.

그러나 게임 결과를 출력한다는 의미의 `ResultView` 와 같은 객체라면 domain 이 아닌 view 패키지에 포함시켜야한다고 생각합니다.
어떻게 출력할 것인지 정보를 담고 있으므로(문자열, 이미지, 웹뷰 등등) 우리 비즈니스, 요구사항과 크게 관련이 없는,
우리 애플리케이션을 구성하는 방법 중 하나를 담은 코드거든요.

nooblette replied:
아 `RaceService`에서 경주 결과를 바로 출력하는게 아니라, 게임 결과를 담는 중간계층(e.g. `GameResult`)를 추가한다면 출력 정보와 출력 방법을 유연하게 분리할 수 있겠네요!

이때 출력 정보인 게임 결과를 비즈니스와 관련된 부분이니, `domain` 로직에 담는게 적절하겠군요!
제가 다소 좁은 시각으로 생각했던 것 같아요 😅 마지막까지 많은 도움이되는 리뷰 주셔서 감사합니다 👍

Hyeon9mak reviewed (approved):
안녕하세요 민혁님~ 자동차 경주 미션 마지막까지 고생 많으셨습니다.
그간 남겨드린 코멘트들이 도움 되셨을까요?

질문 주신 내용에 대한 간단한 답변과 함께 approve 하도록 하겠습니다.
다음 미션들도 화이팅입니다!
